pkg/agent/config/config.go

```diff
@@ -405,14 +406,25 @@ func get(ctx context.Context, envInfo *cmds.Agent, proxy proxy.Proxy) (*config.N
 			return nil, err
 		}
 
+		// Pass ipv4, ipv6 or both depending on nodeIPs mode
 		var vpnIPs []net.IP
-		if vpnInfo.IPv4Address != nil {
-			vpnIPs = append(vpnIPs, vpnInfo.IPv4Address)
+		dualNode, err := utilsnet.IsDualStackIPs(nodeIPs)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to validate node-ip: %v", nodeIPs)
 		}
-		if vpnInfo.IPv6Address != nil {
-			vpnIPs = append(vpnIPs, vpnInfo.IPv6Address)
+		if dualNode && vpnInfo.IPv4Address != nil && vpnInfo.IPv6Address != nil {
+			vpnIPs = append(vpnIPs, vpnInfo.IPv4Address, vpnInfo.IPv6Address)
+		} else {
+			if utilsnet.IsIPv4(nodeIPs[0]) && vpnInfo.IPv4Address != nil {
+				vpnIPs = append(vpnIPs, vpnInfo.IPv4Address)
+			} else if utilsnet.IsIPv6(nodeIPs[0]) && vpnInfo.IPv6Address != nil {
+				vpnIPs = append(vpnIPs, vpnInfo.IPv6Address)
+			} else {
+				return nil, errors.Errorf("failed to replace the old node-ips: %v with the vpn ips: %v, %v", nodeIPs, vpnInfo.IPv4Address, vpnInfo.IPv6Address)
```

Comment:
Not sure if this will ever be seen, but I have a nit on the error message:
```suggestion
				return nil, errors.Errorf("address family mismatch when assigning VPN addresses to node: node=%v, VPN ipv4=%v ipv6=%v", nodeIPs, vpnInfo.IPv4Address, vpnInfo.IPv6Address)
```

Reply:
Yes, it could be seen if users select headscale instead of the official tailscale server and misconfigure things. Thanks, your message is clearer :)